### PR TITLE
toRasterAsync now returns Future.

### PR DESCRIPTION
**Pull request by emilwestergren**
@marcusnaslund @sebastianbaginski what do you think about this?

### CODE REVIEW

File: source/draw/gpu/opengl/OpenGLContext.ooc [hunk 1]
```diff
@@ -164,6 +201,7 @@ OpenGLContext: class extends GpuContext {
 		target canvas draw(source, map)
 	}
 	createFence: override func -> GpuFence { OpenGLFence new(this) }
+	toRasterAsync: override func (source: GpuImage) -> Future<RasterImage> { _DummyFuture new(this toRaster(source)) }
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: The default implementation will actually be synchronized, so we wait on a dummy future.

File: test/draw/gpu/ToRasterTest.ooc [hunk 2]
```diff
@@ -40,15 +40,31 @@ ToRasterTest: class extends Fixture {
 		gpuImage free()
 		raster free()
 	}
+	toRasterAsync: func (sourceImage: RasterImage) {
+		gpuImage := context createImage(sourceImage)
+		future := gpuImage toRasterAsync() as _ToRasterFuture
```
Comment (emilwestergren) [inline test/draw/gpu/ToRasterTest.ooc]: This cast is necessary, otherwise we always get `null` as result.

**Comment by marcusnaslund** [comment]
Looks good to me. `_DummyFuture` is too generic a name when specialised for a certain type, though.

**Comment by emilwestergren** [comment]
Sure, maybe `_DummyToRasterFuture`

**Comment by marcusnaslund** [comment]
That's good.

**Comment by emilwestergren** [comment]
Do you prefer if `toRasterAsync` returns `ToRasterFuture` instead, and we change the API when the ooc bug is fixed?

**Comment by marcusnaslund** [comment]
It's up to you if you want to be explicit about the type everywhere or not. I don't know how soon we will be able to fix this particular bug/design feature, so perhaps don't base the API on it.

**Comment by emilwestergren** [comment]
Yeah I think it's better to define the API as we want it, and go around the bug until it has been fixed.

**Comment by sebastianbaginski** [comment]
I like it because you got rid of `GpuFence` from the API

**Comment by emilwestergren** [comment]
Yes, after this we will remove `GpuFence` and `OpenGLFence` :smiley:

**Comment by emilwestergren** [comment]
@marcusnaslund peer review

**Comment by marcusnaslund** [comment]
Looks good, but:
```
source/draw/gpu/GpuContext.ooc:59:1: 
error: control reaches end of non-void function [-Werror=return-type] }
```

**Comment by emilwestergren** [comment]
Updated

File: test/draw/gpu/ToRasterTest.ooc [hunk 3]
```diff
@@ -40,15 +40,31 @@ ToRasterTest: class extends Fixture {
...
+		future wait()
+		raster := future getResult(null)
+		expect(raster != null)
+		distance := raster distance(sourceImage)
```
Comment (marcusnaslund) [inline test/draw/gpu/ToRasterTest.ooc]: You assign `distance` here but never use it on the next line. That's why PRInt fails.

**Comment by emilwestergren** [comment]
I think I've changed my mind. What do you think of `toRaster: func -> ToRasterFuture` instead? This way you don't need to cast it yourself. I don't think the bug in rock will be solved anytime soon anyway.

File: test/draw/gpu/ToRasterTest.ooc (hunk 3 above)
Comment (emilwestergren) [inline test/draw/gpu/ToRasterTest.ooc]: Oh I see

**Comment by emilwestergren** [comment]
Updated.

**Comment by marcusnaslund** [comment]
Memory leaks.

File: source/draw/gpu/opengl/AndroidContext.ooc [hunk 4]
```diff
@@ -116,25 +116,24 @@ AndroidContext: class extends OpenGLContext {
...
+	toRasterAsync: func ~monochrome (gpuImage: OpenGLMonochrome) -> ToRasterFuture {
 		(buffer, fence) := this toBuffer(gpuImage, this _packMonochrome)
-		(RasterMonochrome new(buffer, gpuImage size), fence)
+		_FenceToRasterFuture new(RasterMonochrome new(buffer, gpuImage size), fence)
 	}
-	toRasterAsync: func ~uv (gpuImage: OpenGLUv) -> (RasterImage, GpuFence) {
+	toRasterAsync: func ~uv (gpuImage: OpenGLUv) -> ToRasterFuture {
 		(buffer, fence) := this toBuffer(gpuImage, this _packUv)
-		(RasterUv new(buffer, gpuImage size), fence)
+		_FenceToRasterFuture new(RasterUv new(buffer, gpuImage size), fence)
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/AndroidContext.ooc]: Note that the return type is `ToRasterFuture`. When you call `free` on whatever this method returns, and don't explicitly cast (I can't see any explicit cast), you will call `ToRasterFuture::free` which doesn't free the `GpuFence` instance.

Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: What? Why? Because of the rock bug?

Comment (sebastianbaginski) [inline source/draw/gpu/opengl/AndroidContext.ooc]: That's so weird.

Comment (marcusnaslund) [inline source/draw/gpu/opengl/AndroidContext.ooc]: It's the same thing as before, the type is inferred as the specified function return type, not as the object you actually return.

Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: But `free` is a virtual function, it shouldn't depend on the returned type...

Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: Even `getResult` actually calls the correct function (i.e. `getResult` in the implemented type), it's just the returned value that's corrupt.

Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: I think the memory leak is due to my adaption to this change in Vidproc.

Comment (marcusnaslund) [inline source/draw/gpu/opengl/AndroidContext.ooc]: Oh, alright then. When it comes to rock, I usually assume the worst.

Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: Sounds like a fair strategy ;)

**Comment by marcusnaslund** [comment]
#908 specifies that `ToRasterAsync` should be renamed, will that be in the next PR?

**Comment by emilwestergren** [comment]
Yes, that will be a separate PR. It's a bit more messy since a lot of tests need to be adapted.

**Pull request merged**